File: planet_finder.js
#!/usr/bin/env node
// planet_finder.js
const { program } = require('commander');
const chalk = require('chalk');
const fs = require('fs');

// Данные планет (средние элементы)
const PLANET_DATA = {
    Mercury: { L0: 252.2509, n: 4.09233445 },
    Venus:   { L0: 181.9798, n: 1.60213043 },
    Earth:   { L0: 100.4664, n: 0.98560911 },
    Mars:    { L0: 355.4530, n: 0.52407178 },
    Jupiter: { L0: 34.3515,  n: 0.08309125 },
    Saturn:  { L0: 50.0774,  n: 0.03344482 },
    Uranus:  { L0: 314.0550, n: 0.01172312 },
    Neptune: { L0: 304.3487, n: 0.00597733 },
};
const OBLIQUITY = 23.4393; // градусы

function eclipticToEquatorial(lon, lat = 0, obl = OBLIQUITY) {
    const lonRad = lon * Math.PI / 180;
    const latRad = lat * Math.PI / 180;
    const oblRad = obl * Math.PI / 180;
    const ra = Math.atan2(Math.sin(lonRad) * Math.cos(oblRad) - Math.tan(latRad) * Math.sin(oblRad),
                          Math.cos(lonRad));
    const dec = Math.asin(Math.sin(latRad) * Math.cos(oblRad) + Math.cos(latRad) * Math.sin(oblRad) * Math.sin(lonRad));
    return { ra: (ra * 180 / Math.PI) % 360, dec: dec * 180 / Math.PI };
}

function raToHms(raDeg) {
    const hours = raDeg / 15;
    const h = Math.floor(hours);
    const m = Math.floor((hours - h) * 60);
    const s = (hours - h - m/60) * 3600;
    return `${String(h).padStart(2,'0')}h ${String(m).padStart(2,'0')}m ${s.toFixed(2)}s`;
}

function decToDms(decDeg) {
    const sign = decDeg >= 0 ? '+' : '-';
    decDeg = Math.abs(decDeg);
    const d = Math.floor(decDeg);
    const m = Math.floor((decDeg - d) * 60);
    const s = (decDeg - d - m/60) * 3600;
    return `${sign}${String(d).padStart(2,'0')}° ${String(m).padStart(2,'0')}' ${s.toFixed(2)}"`;
}

function computePlanetPosition(date, planet) {
    const j2000 = new Date(Date.UTC(2000, 0, 1, 12, 0));
    const delta = (date - j2000) / (1000 * 60 * 60 * 24);
    const data = PLANET_DATA[planet];
    const lon = (data.L0 + data.n * delta) % 360;
    const { ra, dec } = eclipticToEquatorial(lon);
    return { ra, dec };
}

program
    .option('--date <date>', 'Дата (YYYY-MM-DD)')
    .option('--planet <name>', 'Название планеты')
    .option('--list', 'Показать все планеты')
    .option('--export-json <file>', 'Экспорт в JSON')
    .option('--export-csv <file>', 'Экспорт в CSV')
    .parse(process.argv);

const opts = program.opts();

let date;
if (opts.date) {
    date = new Date(opts.date + 'T00:00:00Z');
    if (isNaN(date)) {
        console.error(chalk.red('Неверный формат даты. Используйте YYYY-MM-DD'));
        process.exit(1);
    }
} else {
    date = new Date();
}

let planets;
if (opts.planet) {
    if (!PLANET_DATA[opts.planet]) {
        console.error(chalk.red(`Планета '${opts.planet}' не найдена.`));
        process.exit(1);
    }
    planets = [opts.planet];
} else {
    planets = Object.keys(PLANET_DATA);
}

const results = [];
for (const p of planets) {
    const { ra, dec } = computePlanetPosition(date, p);
    results.push({
        planet: p,
        ra_deg: ra,
        dec_deg: dec,
        ra: raToHms(ra),
        dec: decToDms(dec)
    });
}

if (opts.exportJson) {
    fs.writeFileSync(opts.exportJson, JSON.stringify(results, null, 2));
    console.log(chalk.green(`Экспортировано в ${opts.exportJson} (JSON)`));
} else if (opts.exportCsv) {
    const header = 'planet,ra_deg,dec_deg,ra_hms,dec_dms\n';
    const rows = results.map(r => `${r.planet},${r.ra_deg},${r.dec_deg},${r.ra},${r.dec}`).join('\n');
    fs.writeFileSync(opts.exportCsv, header + rows);
    console.log(chalk.green(`Экспортировано в ${opts.exportCsv} (CSV)`));
} else {
    console.log(chalk.cyan(`Положение планет на ${date.toISOString().split('T')[0]}:\n`));
    for (const r of results) {
        console.log(chalk.yellow(`${r.planet}:`));
        console.log(`  RA  = ${r.ra}  (${r.ra_deg.toFixed(2)}°)`);
        console.log(`  Dec = ${r.dec}  (${r.dec_deg.toFixed(2)}°)`);
        console.log();
    }
}

File: planet_finder.go
// planet_finder.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

type PlanetData struct {
	L0 float64
	N  float64
}

var planetData = map[string]PlanetData{
	"Mercury": {252.2509, 4.09233445},
	"Venus":   {181.9798, 1.60213043},
	"Earth":   {100.4664, 0.98560911},
	"Mars":    {355.4530, 0.52407178},
	"Jupiter": {34.3515, 0.08309125},
	"Saturn":  {50.0774, 0.03344482},
	"Uranus":  {314.0550, 0.01172312},
	"Neptune": {304.3487, 0.00597733},
}
const OBLIQUITY = 23.4393 // degrees

type Result struct {
	Planet  string  `json:"planet"`
	RADeg   float64 `json:"ra_deg"`
	DecDeg  float64 `json:"dec_deg"`
	RA      string  `json:"ra"`
	Dec     string  `json:"dec"`
}

func eclipticToEquatorial(lon, lat, obl float64) (float64, float64) {
	lonRad := lon * math.Pi / 180
	latRad := lat * math.Pi / 180
	oblRad := obl * math.Pi / 180
	ra := math.Atan2(math.Sin(lonRad)*math.Cos(oblRad)-math.Tan(latRad)*math.Sin(oblRad),
		math.Cos(lonRad))
	dec := math.Asin(math.Sin(latRad)*math.Cos(oblRad) + math.Cos(latRad)*math.Sin(oblRad)*math.Sin(lonRad))
	raDeg := math.Mod(ra*180/math.Pi+360, 360)
	decDeg := dec * 180 / math.Pi
	return raDeg, decDeg
}

func raToHms(raDeg float64) string {
	hours := raDeg / 15
	h := int(hours)
	m := int((hours - float64(h)) * 60)
	s := (hours - float64(h) - float64(m)/60) * 3600
	return fmt.Sprintf("%02dh %02dm %05.2fs", h, m, s)
}

func decToDms(decDeg float64) string {
	sign := '+'
	if decDeg < 0 {
		sign = '-'
		decDeg = -decDeg
	}
	d := int(decDeg)
	m := int((decDeg - float64(d)) * 60)
	s := (decDeg - float64(d) - float64(m)/60) * 3600
	return fmt.Sprintf("%c%02d° %02d' %05.2f\"", sign, d, m, s)
}

func computePlanetPosition(date time.Time, planet string) (float64, float64) {
	j2000 := time.Date(2000, 1, 1, 12, 0, 0, 0, time.UTC)
	delta := date.Sub(j2000).Hours() / 24
	data := planetData[planet]
	lon := math.Mod(data.L0+data.N*delta+360, 360)
	ra, dec := eclipticToEquatorial(lon, 0, OBLIQUITY)
	return ra, dec
}

func main() {
	var (
		dateStr    string
		planetName string
		list       bool
		exportJson string
		exportCsv  string
	)
	flag.StringVar(&dateStr, "date", "", "Дата (YYYY-MM-DD)")
	flag.StringVar(&planetName, "planet", "", "Название планеты")
	flag.BoolVar(&list, "list", false, "Показать все планеты")
	flag.StringVar(&exportJson, "export-json", "", "Экспорт в JSON")
	flag.StringVar(&exportCsv, "export-csv", "", "Экспорт в CSV")
	flag.Parse()

	var date time.Time
	if dateStr != "" {
		var err error
		date, err = time.Parse("2006-01-02", dateStr)
		if err != nil {
			fmt.Println("\033[31mНеверный формат даты. Используйте YYYY-MM-DD\033[0m")
			os.Exit(1)
		}
	} else {
		date = time.Now()
	}

	var planets []string
	if planetName != "" {
		if _, ok := planetData[planetName]; !ok {
			fmt.Printf("\033[31mПланета '%s' не найдена.\033[0m\n", planetName)
			os.Exit(1)
		}
		planets = []string{planetName}
	} else {
		for p := range planetData {
			planets = append(planets, p)
		}
	}

	results := make([]Result, 0, len(planets))
	for _, p := range planets {
		ra, dec := computePlanetPosition(date, p)
		results = append(results, Result{
			Planet:  p,
			RADeg:   ra,
			DecDeg:  dec,
			RA:      raToHms(ra),
			Dec:     decToDms(dec),
		})
	}

	if exportJson != "" {
		data, _ := json.MarshalIndent(results, "", "  ")
		os.WriteFile(exportJson, data, 0644)
		fmt.Printf("\033[32mЭкспортировано в %s (JSON)\033[0m\n", exportJson)
	} else if exportCsv != "" {
		f, _ := os.Create(exportCsv)
		defer f.Close()
		w := csv.NewWriter(f)
		defer w.Flush()
		w.Write([]string{"planet", "ra_deg", "dec_deg", "ra_hms", "dec_dms"})
		for _, r := range results {
			w.Write([]string{r.Planet, fmt.Sprintf("%f", r.RADeg), fmt.Sprintf("%f", r.DecDeg), r.RA, r.Dec})
		}
		fmt.Printf("\033[32mЭкспортировано в %s (CSV)\033[0m\n", exportCsv)
	} else {
		fmt.Printf("\033[36mПоложение планет на %s:\033[0m\n\n", date.Format("2006-01-02"))
		for _, r := range results {
			fmt.Printf("\033[33m%s:\033[0m\n", r.Planet)
			fmt.Printf("  RA  = %s  (%.2f°)\n", r.RA, r.RADeg)
			fmt.Printf("  Dec = %s  (%.2f°)\n", r.Dec, r.DecDeg)
			fmt.Println()
		}
	}
}
